### PR TITLE
documents important MC vars and procs and adds tracking of tick usage when the mc first wakes up and other stats

### PR DESCRIPTION
**Pull request by Kylerace**
## About The Pull Request
tick diagram2
this was part of an effort to further understand the MC and help quantify how much time it has to actually run within the average tick. other sleeping procs can resume before the MC does which pushes it out of the tick and we dont seem to track that, so i added that with average_starting_tick_usage which is a running average the tick usage of when the MC starts running for the tick. 

Screenshot_1668
this is what the current new Master entry looks like. It was split up into Master and Tick Division entries.

- TickRate has been renamed to Ticks Per Iteration since thats a less ambiguous name for the processing var
- Max Tick Limit Im still not sure what to do with since im 99% sure the one you see in the stat entry is whatever the running tick limit was when SSstat_panels was fire()'d which means it doesnt have a meaning relevant to master as a whole
- Sleeping Overtime im not sure if ill keep but it is how much sleeping procs resumed after the MC start after the scheduled end of the tick that master fired in.
- stoplag Threads is how many concurrent stoplag()'s there are - might be useful in showing tick contention on live i intend to experiment with it (note from live testing: debatably useful)
- Ticks Skipped is a running average of how many ticks the MC goes to sleep for minus 1 (because sleeping for one tick is normal, it just makes it resume on the next tick). this is a stat otherwise completely invisible to the stat panel but leads to entire missed ticks of potential processing time so i think this one is pretty good

The Tick Division entry shows measurements of what each portion of the tick took up. Meant to show whats eating up the average tick beyond maptick and the MC itself.
- Pre MC is a running average of the tick usage when master's main loop resumes. If this is >0 it means there are sleeping procs resuming before master which can cause master to not process if it starts too late
- MC shows the average tick usage of each master controller run (ending tick usage minus starting tick usage)
- Post MC is how much tick usage sleeping procs take after the MC has gone to sleep but before SendMaps processes. as predicted by mso's old comments this is much less common than pre MC usage because most sleeps tend to be longer than the MC's sleeps and sleeps are ran in the order theyre called.
- Maptick has been moved into this category (and renamed from Internal Tick Rate)
- Post Maptick is an attempt to measure the tick usage of client verbs/procs executing after SendMaps processes. this one is likely the least accurate since 1. the measurements are only updated for verbs/procs that have a define that sets a tick usage var to the ending tick usage so not all of them actually update it and 2. it has to use world.map_cpu to determine that portions tick usage, which is an average of the last x ticks instead of measuring a single tick. However I think it could be a useful measurement we'll see

Screenshot_1651
also added an additional stat on every subsystems stat_entry which is the percentage of the average tick this subsystem takes up whenever it ignites, added at the end of the default stat_entry message. Its essentially a running average of time spent running / time spent not running. So in this example lighting takes 0.1% of all recent processing and machines takes 1%. This is important because it makes it easier to see what time a subsystem takes from the server as opposed to most other stats which show hwo long a given subsystem takes to complete its arbitrary run of processing

## Why It's Good For The Game
In general, works to quantify and log more information about what goes on within a tick.

also documents and clarifies some things about the MC

### CODE REVIEW

File: code/controllers/master.dm [hunk 1]
```diff
@@ -342,6 +352,9 @@ GLOBAL_REAL(Master, /datum/controller/master) = new
 	while (1)
 		tickdrift = max(0, MC_AVERAGE_FAST(tickdrift, (((REALTIMEOFDAY - init_timeofday) - (world.time - init_time)) / world.tick_lag)))
 		var/starting_tick_usage = TICK_USAGE
+
+		average_starting_tick_usage = MC_AVG_FAST_UP_SLOW_DOWN(average_starting_tick_usage, starting_tick_usage)
```
Comment (LemonInTheDark) [inline code/controllers/master.dm:356]: could you log this in with td and maptick? I want to graph

Comment (Kylerace) [reply code/controllers/master.dm:356]: how could i forget

Comment (LemonInTheDark) [reply code/controllers/master.dm:356]: thx xoxo

File: code/controllers/master.dm [hunk 2]
```diff
@@ -55,13 +68,19 @@ GLOBAL_REAL(Master, /datum/controller/master) = new
...
+	///!End of queue linked list (used for appending to the list)
+	var/datum/controller/subsystem/queue_tail
+	///Running total of all queued subsystem priorities so that we don't have to loop thru the queue each run to split up the tick
+	var/queue_priority_count = 0
+	///running total of queued subsystem priority but for background subsystems
+	var/queue_priority_count_bg = 0
+	///!Are we loading in a new map?
+	var/map_loading = FALSE
+
+	///!for scheduling different subsystems for different stages of the round
```
Comment (MrStonedOne) [inline code/controllers/master.dm:82]: if you're gonna move these to before the thing, you have to remove the `!` from the comment as that tells codedocs to attach the comment to the *previous* thing, not the next thing.

File: code/controllers/master.dm [hunk 3]
```diff
@@ -289,34 +308,33 @@ GLOBAL_REAL(Master, /datum/controller/master) = new
...
+		if (subsystem.flags & SS_NO_FIRE)
 			continue
-		SS.queued_time = 0
-		SS.queue_next = null
-		SS.queue_prev = null
-		SS.state = SS_IDLE
-		if (SS.flags & SS_TICKER)
-			tickersubsystems += SS
+		subsystem.queued_time = 0
+		subsystem.queue_next = null
```
Comment (MrStonedOne) [inline code/controllers/master.dm:315]: You have been docked 420 social credit score points.

Comment (Kylerace) [reply code/controllers/master.dm:315]: image

File: code/controllers/master.dm (hunk 2 above)
Comment (Kylerace) [reply code/controllers/master.dm:82]: ah right

**Review by MrsTonedOne — COMMENTED** [review]
@github-actions github-actions bot requested a review from MrStonedOne 3 minutes ago

File: code/controllers/master.dm (hunk 2 above)
Comment (Kylerace) [reply code/controllers/master.dm]: fixed

File: code/controllers/master.dm [hunk 4]
```diff
@@ -615,24 +657,23 @@ GLOBAL_REAL(Master, /datum/controller/master) = new
 
 
 /datum/controller/master/stat_entry(msg)
-	msg = "(TickRate:[Master.processing]) (Iteration:[Master.iteration]) (TickLimit: [round(Master.current_ticklimit, 0.1)])"
+	msg = "(TickRate:[Master.processing]) (Iteration:[Master.iteration]) (TickLimit: [round(Master.current_ticklimit, 0.1)]) (Average Starting Tick Usage: [round(Master.average_starting_tick_usage, 0.1)] \
+	(Average Sleeping TIck Usage: [round(Master.average_sleeping_tick_usage, 0.1)]) (Average Sleeping Overtime: [round(Master.average_sleeping_overtime_usage)]))"
```
Comment (LemonInTheDark) [inline code/controllers/master.dm:661]: Do we need the average here? more similar words there are the harder it is to read at a glance

File: code/controllers/subsystem/time_track.dm [hunk 5]
```diff
@@ -59,6 +59,9 @@ SUBSYSTEM_DEF(time_track)
 			"tidi_avg",
 			"tidi_slowavg",
 			"maptick",
+			"average starting tick usage",
+			"average sleeping tick usage",
+			"average sleeping tick overtime",
```
Comment (LemonInTheDark) [inline code/controllers/subsystem/time_track.dm:64]: _ seperated please, also avg_start_tick_use maybe? or just sleep_use_avg and sleep_over_avg? less longwinded the better, while still remaining clear and all

File: code/controllers/master.dm (hunk 4 above)
Comment (Kylerace) [reply code/controllers/master.dm:661]: i shortened it to Avg ill see if that makes it better as i adjust it more (and add more stats)

File: code/__DEFINES/MC.dm [hunk 6]
```diff
@@ -79,3 +79,14 @@
...
+
+///RunQueue() didnt complete a run and something fucked up
+#define RUNQUEUE_ERRORED FALSE
+///RunQueue() successfully completed a run
+#define RUNQUEUE_SUCCESSFUL_RUN TRUE
+
+///CheckQueue() wasnt able to go through its list of subsystems and something fucked up
+#define CHECKQUEUE_ERRORED FALSE
+///CheckQueue() was able to go through its list of subsystems
+#define CHECKQUEUE_SUCCESSFUL_RUN TRUE
```
Comment (LemonInTheDark) [inline code/__DEFINES/MC.dm:91]: Why not just TRUE/FALSE?

Comment (Kylerace) [reply code/__DEFINES/MC.dm:91]: eh probably better to keep it to TRUE/FALSE i just thought they were important enough to warrant the define

Comment (LemonInTheDark) [reply code/__DEFINES/MC.dm:91]: It's the same pattern as most other procs that succeed/fail right?

**Comment by Kylerace** [comment]
average_percentage_of_tick is inaccurate in some way, i suspect its when a subsystem frequently pauses during a fire and resumes the next tick

**Comment by Kylerace** [comment]
doing other stuff ill reopen this later

**Comment by LemonInTheDark** [comment]
you had better

**Pull request closed without merging**